crates/sui-core/src/authority_active/gossip/node_sync.rs

```diff
@@ -0,0 +1,435 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use super::{DigestHandler, Follower};
+use crate::{
+    authority::AuthorityState, authority_aggregator::AuthorityAggregator,
+    authority_client::AuthorityAPI,
+};
+use async_trait::async_trait;
+
+use std::collections::{hash_map, HashMap};
+use sui_storage::node_sync_store::NodeSyncStore;
+use sui_types::{
+    base_types::{AuthorityName, ExecutionDigests, TransactionDigest},
+    committee::{Committee, StakeUnit},
+    error::{SuiError, SuiResult},
+    messages::{CertifiedTransaction, ConfirmationTransaction, SignedTransactionEffects},
+};
+
+use std::ops::Deref;
+use std::sync::Arc;
+
+use tokio::sync::{broadcast, mpsc, Mutex, Semaphore};
+use tokio::task::JoinHandle;
+
+use tracing::{error, info, trace, warn};
+
+const NODE_SYNC_QUEUE_LEN: usize = 500;
+
+// Process up to 20 digests concurrently.
+const MAX_NODE_SYNC_CONCURRENCY: usize = 100;
+
+/// EffectsStakeMap tracks which effects digests have been attested by a quorum of validators and
+/// are thus final.
+struct EffectsStakeMap {
+    /// Keep track of how much stake has voted for a given effects digest
+    /// any entry in this map with >2f+1 stake can be sequenced locally and
+    /// removed from the map.
+    effects_stake_map: HashMap<ExecutionDigests, StakeUnit>,
+    /// Keep track of stake votes per validator - needed to double check the total stored in
+    /// effects_stake_map, which can otherwise be corrupted by byzantine double-voting.
+    effects_vote_map: HashMap<(ExecutionDigests, AuthorityName), StakeUnit>,
+}
+
+impl EffectsStakeMap {
+    pub fn new() -> Self {
+        Self {
+            effects_stake_map: HashMap::new(),
+            effects_vote_map: HashMap::new(),
+        }
+    }
+
+    /// Note that a given effects digest has been attested by a validator, and return true if the
+    /// stake that has attested that effects digest has exceeded the quorum threshold.
+    pub fn note_effects_digest(
+        &mut self,
+        source: &AuthorityName,
+        stake: StakeUnit,
+        quorum_threshold: StakeUnit,
+        digests: &ExecutionDigests,
+    ) -> bool {
+        let vote_entry = self.effects_vote_map.entry((*digests, *source));
+
+        let cur_stake = if let hash_map::Entry::Occupied(_) = &vote_entry {
+            // TODO: report byzantine authority suspciion
+            warn!(peer = ?source, digest = ?digests.transaction, effects = ?digests.effects,
+                "ByzantineAuthoritySuspicion: peer double-voted for effects digest");
```

Comment:
I'm not exactly sure what you mean by "double vote". Nothing in the protocol lets you vote twice? Or is this taking care of the fact that a validator literally puts their sig twice on an effects?

Reply:
Seeing it a second time shouldn't be a problem though

Reply:
It is if you add in that validators stake every time you see it! Which is what I'm preventing here.